### ISSUE TITLE
linen api:  no check on  variables and param shapes coherence

### ISSUE THREAD

**Issue by mortner31** [google/flax#477]
### Problem you have encountered:

When performing transfer of params to a module it is possible to feed the module with params that do not have proper shape

```python
from flax import linen as nn
from jax.config import config
from jax import random

config.enable_omnistaging()

key1, key2 = random.split(random.PRNGKey(0), 2)
x = random.uniform(key1, (1,3,3,1))

conv_module_5 =nn.Conv(features=5,kernel_size=(3,3))
param_features_5 = conv_module_5.init(key2, x)

conv_module_3 =nn.Conv(features=3,kernel_size=(3,3))

 # this should raise an error
y = conv_module_3.apply(param_features_5, x)

assert y.shape[3] == 3, "module output has improper shape"

```
### What you expected to happen:

I would expect an error while feeding the `conv_module_3` with improper shape. 

### Steps to reproduce:

Here is a colab

https://colab.research.google.com/drive/1Y7WOXjxgkYPldyzPQBcP5PdtEY6EPEAX?usp=sharing

**Comment by avital** [comment]
Thanks for the report @mortner31 -- this is due to the logic in https://github.com/google/flax/blob/76d580dbc157ac8065be46665e17790c74f92ed5/flax/linen/module.py#L422, where we don't assume any special treatment for the additional arguments passed to the initializer function (other than the RNG).

Current Flax actually has the check that you're asking for, by hardcoding that the first argument is a shape. That harms generality a bit, and was even discussed for current Flax in #160 -- note that in that issue:
> One thing is that we'd lose the shape checking in Module.params, but that seems like the kind of check which should be part of a test anyway.

(I'm not sure I agree with the point above -- being realistic, much research code doesn't have tests and it's unfortunate to require tests to catch issues that could be caught earlier)

So now I'm no longer sure I support #160, and perhaps you're right and we should special case the first argument to `self.param` (and probably `self.variable` at that point?) to be a shape, and bring back the shape equality check.

But should we do that by re-adding a `shape` argument explicitly? Or by inspecting the first argument of `*init_args` (in case there are arguments?). I'm not totally sure.

Another possibility, which I'm not sure entirely about, is to add an `else` clause here: https://github.com/google/flax/blob/76d580dbc157ac8065be46665e17790c74f92ed5/flax/core/scope.py#L255, to say -- even if the variable is present, run the initializer function with a fake PRNG, just to see that you get the same shape. IIUC, because shapes are known during tracing, that code will not be present in the compiled JAXPR, meaning the accelerator code won't be impacted. But compilation will become a bit slower. We could also consider something like this in "DEBUG" mode.

Roping in @jheek who thought a lot about this when building current Flax.

**Comment by avital** [comment]
cc @levskaya for opinions as well.

**Comment by jheek** [comment]
Actually I think we can make the check more powerful than the original one. Before we forced a static shape argument and checked that with the shape of the type. This had a number of issues:
1. The shape could not be determined by the init function
2. Shape checks could not be more complex than a single array. So no possibility shape check a pytree for example.


What we can do instead is use `jax.eval_shape` to check whether the initializer would still return a result with the same shape and dtype. The only complexity here is that we should make sure the check doesn't actually require or split PRNG sequences.

**Comment by avital** [comment]
@jheek -- that sounds pretty good! Is this something you're working on?

**Comment by avital** [comment]
This is also related to #86

**Comment by avital** [comment]
This is also related to #160

**Comment by avital** [comment]
Oh, I believe this was resolved in #530
